botocore/signers.py

```diff
@@ -196,6 +208,65 @@ def generate_presigned_url(self, request_dict, expires_in=3600,
         return request.url
 
 
+class CloudFrontSigner(object):
+
+    @staticmethod
+    def b64encode(content):
+        return base64.b64encode(content).translate(maketrans(b'+=/', b'-_~'))
+
+    def __init__(self, key_id, private_key):
+        """Create a CloudFrontSigner.
+
+        :type key_id: str
+        :param key_id: The CloudFront Key Pair ID
+
+        :type private_key: str
+        :param private_key: The private key string used for signing.
+        """
+        self.key_id = key_id
+        self.rsa_pk = rsa.PrivateKey.load_pkcs1(private_key.encode('utf8'))
+
+    def sign(self, url, expires, starts=None, ip_address=None):
+        """Sign a given url.
+
+        :type url: str
+        :param url: The URL of the protected object.
+
+        :type expires: int
+        :param expires: The expiry time of the URL. Format is a timestamp.
+
+        :type starts: int
+        :param starts: The start time of the URL. Format is a timestamp.
+
+        :type ip_address: str
+        :param ip_address: Use 'x.x.x.x' for an IP, or 'x.x.x.x/x' for a subnet
```

Comment:
Thanks @mtdowling, I did not know you implemented the CloudFront in PHP. Good job! We could have talked more about this topic beforehand.

You have several similar comments in this page. Please allow me to respond them here in one place. Yes I am aware of the "canned policy" and "custom policy" difference. In my understanding, the name "custom policy" is misleading in the first place. Customer can not really customize an arbitrary policy (just like what they can do in IAM policies), and expects the API to just sign it. [CloudFront only accepts "custom policy" with combinations of predefined 3 parameters in one format](http://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/private-content-creating-signed-url-custom-policy.html#private-content-custom-policy-statement). So I can not see a scenario that customer would need more flexibility to build the policy by themselves and call a separated sign_custom_policy() function.

In other words, despite of the subtle output difference of so-called "canned policy" and "custom policy", their input are all driven by a subset of same parameters. There won't be a need to expose a public sign_custom_policy() for a "customized" policy.